feat: issue_02

yoshikawa-akr opened the pull request:
# [issue No.2](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#2%E3%81%93%E3%81%AEdom%E3%82%92javascript%E3%81%A7%E3%81%A4%E3%81%8F%E3%82%8Ahtml%E5%86%85%E3%81%AEul%E3%81%AE%E4%B8%AD%E3%81%AB%E5%B7%AE%E3%81%97%E8%BE%BC%E3%82%93%E3%81%A7%E3%81%8F%E3%81%A0%E3%81%95%E3%81%84)

**issue**
> このDOMをJavaScriptでつくり、html内のulの中に差し込んでください
``` 
<li>
  <a href="1.html"><img src="bookmark.png" alt="ブックマーク" />これです</a>
</li>
```
## [CodeSandBox](https://codesandbox.io/s/twilight-darkness-cjzq9) / Latest commit: https://github.com/yoshikawa-akr/js-lessons/pull/2/commits/7153bf1f4d316bb3963d19b4b50d414311a7a8b6

## In terms of implementation, what you are worried about, what you want to focus on, and what you want to ask.
I've written a number of processes using appendChild, and I was wondering if it was possible to process them all together.
- 943b6b75b1d7022cbe0f53f7297f165a6939a235
- 672e9a86cb0978390878d1f41bbd74429cba4afc
- 3c1ff00ff0557d0df22f663c9323d2e5dceb36f0

yoshikawa-akr committed:
feat: Getting the ul element.

Give an id to the ul element and get the element.

yoshikawa-akr committed:
refactor: Deleting and modifying code.

Change lang attribute, change unnecessary code.

yoshikawa-akr committed:
feat: Creating the a tag.

Create an a tag and specify the href attribute.

yoshikawa-akr committed:
feat: Creating an img tag.

Create an img tag and specify the src and alt.

yoshikawa-akr committed:
refactor: Add wording to a tag.

yoshikawa-akr committed:
feat: Add the img tag to the child element of the a tag.

yoshikawa-akr committed:
feat: Creating a li tag.

Create a li tag and add an a tag as a child element.

yoshikawa-akr committed:
feat: Specify li tag as a child element of ul tag.

443n0511 commented on issue_02/script.js:
```diff
@@ -0,0 +1,11 @@
+const img = document.createElement('img');
+img.src = 'bookmark.png';
+img.alt = 'ブックマーク';
+a.appendChild(img);
+const li = document.createElement('li');
+li.appendChild(a);
+ul.appendChild(li);
```
MUST
### The first.
```js
<li>
  <a href="1.html"><img src="bookmark.png" alt="ブックマーク" />これです</a>
</li>
```
The specification specifies that the output should look like this, but in sandbox it looks like this

<img width="319" alt="スクリーンショット 2022-02-12 20 56 45" src="https://user-images.githubusercontent.com/43788628/153710385-d0526045-5da9-4f11-929c-cf560fb5f31e.png">

Think about where the img element should be inserted and check it out!

### The second.
The link to the image is broken in the sandbox. Could you check it for me?

yoshikawa-akr replied:
Thanks for the review!

### The first
First of all, I am very sorry to have taken up your time with my review for my lack of confirmation.
 I will take care not to do this in the future.

Fixed the insertion point of the element.
https://github.com/yoshikawa-akr/js-lessons/pull/2/commits/819492a8bc126a13513d41a916e6f80f782012f0

---
### The second
This one was a mistake on my part.
I thought that setting up the image link itself was the purpose of the assignment, and that there was no need to actually prepare the image file. I apologize for that.

The image file has been prepared and the broken link has been resolved.
https://github.com/yoshikawa-akr/js-lessons/pull/2/commits/c2b9172752c338debc6d99589763c72ccdfc8b2c

443n0511 commented on issue_02/script.js:
```diff
@@ -0,0 +1,11 @@
+img.alt = 'ブックマーク';
+a.appendChild(img);
+const li = document.createElement('li');
+li.appendChild(a);
+ul.appendChild(li);
```
IMO
It can be written in abbreviated form like this.
Do some research on the Mezzot chain.
```suggestion
ul.appendChild(li).appendChild(a);
```

yoshikawa-akr replied:
I rewrote the code to the one you suggested.
https://github.com/yoshikawa-akr/js-lessons/pull/2/commits/297e1347b7f62e22363f973dc81f02e202795f3e

Thanks to your review here, I was able to learn about method chaining. Thank you very much!

yoshikawa-akr committed:
fix: Fix broken links.

yoshikawa-akr committed:
refactor: Change method.

Use insertAdjacentElement to modify the position of img tags.

yoshikawa-akr committed:
refactor: Omit the code.

Use method chains to omit code.

443n0511 commented on issue_02/script.js:
```diff
@@ -0,0 +1,10 @@
+const ul = document.getElementById('js-ul');
+const a = document.createElement('a');
+a.href = '1.html';
+a.textContent = 'これです';
+const img = document.createElement('img');
+img.src = 'bookmark.png';
+img.alt = 'ブックマーク';
+a.insertAdjacentElement('afterbegin', img);
+const li = document.createElement('li');
+ul.appendChild(li).appendChild(a);
```
IMO
This is completely my preference, so I'll leave it to you to decide if you want to fix it or not.
I felt it would be easier to read the code if the constants and the processing related to them were written separately.
```suggestion
const ul = document.getElementById('js-ul');
const li = document.createElement('li');
const a = document.createElement('a');
const img = document.createElement('img');

img.src = 'bookmark.png';
img.alt = 'ブックマーク';
a.href = '1.html';
a.textContent = 'これです';
a.insertAdjacentElement('afterbegin', img);
ul.appendChild(li).appendChild(a);

```

yoshikawa-akr replied:
**Thanks for the comment and approve!**
As you pointed out, I think it is easier to read by separating the variables from the processing for them.
I will try to keep this in mind in the future.

yoshikawa-akr committed:
refactor: Improved readability

The readability was improved by writing the constants and their processing descriptions separately.

hippopo-moo commented on issue_02/script.js:
```diff
@@ -0,0 +1,11 @@
+const a = document.createElement('a');
+const img = document.createElement('img');
+
+img.src = 'bookmark.png';
+img.alt = 'ブックマーク';
+a.href = '1.html';
+a.textContent = 'これです';
```
I don't think there's anything wrong with the code.
I'll share what I learned when I did this assignment before.

When setting attribute values, as Yoshikawa-san writes in these lines, there are two ways to write them: by specifying properties or by using setAttribute.
He says that setAttribute is used when you want to give a custom attribute that the DOM doesn't have.

yoshikawa-akr replied:
**Thank you for your comment and review!**

I've never heard of `setAttribute` before.
According to the [documentation](https://developer.mozilla.org/ja/docs/Web/API/Element/setAttribute), it seems to be a good idea to use it when creating a new attribute.
I'll try to use it in the future when I get the chance!
